ksm.go

```diff
@@ -0,0 +1,497 @@
+	ksmSleepMillisec = "sleep_millisecs"
+	ksmStart         = "1"
+	ksmStop          = "0"
+	defaultKSMMode   = ksmAuto
+)
+
+type sysfsAttribute struct {
+	path string
+	file *os.File
+}
+
+func (attr *sysfsAttribute) open() error {
+	file, err := os.OpenFile(attr.path, os.O_RDWR|syscall.O_NONBLOCK, 0660)
+	attr.file = file
+	return err
+}
+
+func (attr *sysfsAttribute) close() error {
+	err := attr.file.Close()
+	attr.file = nil
+	return err
+}
+
+func (attr *sysfsAttribute) read() (string, error) {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return "", err
+	}
+
+	data, err := ioutil.ReadAll(attr.file)
+	if err != nil {
+		return "", err
+	}
+
+	return string(data), nil
+}
+
+func (attr *sysfsAttribute) write(value string) error {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+
+	err = attr.file.Truncate(0)
+	if err != nil {
+		return err
+	}
+
+	_, err = attr.file.WriteString(value)
+
+	return err
+}
+
+type ksm struct {
+	root          string
+	run           sysfsAttribute
+	pagesToScan   sysfsAttribute
+	sleepInterval sysfsAttribute
+	initialized   bool
+
+	initialPagesToScan   string
+	initialSleepInterval string
+	initialKSMRun        string
+
+	currentKnob ksmMode
+	throttling  bool
+	kickChannel chan bool
+
+	sync.Mutex
+}
+
+func (k *ksm) isAvailable() error {
+	info, err := os.Stat(k.root)
+	if err != nil || !info.IsDir() {
+		return fmt.Errorf("%s is not available", k.root)
+	}
+
+	return nil
+}
+
+func (k *ksm) init(root string) error {
+	var err error
+
+	k.Lock()
+	defer k.Unlock()
+
+	k.initialized = false
+	k.throttling = false
+	k.root = root
+
+	if root == "" {
+		return errors.New("Invalid KSM root")
+	}
+
+	if err := k.isAvailable(); err != nil {
+		return err
+	}
+
+	k.pagesToScan = sysfsAttribute{
+		path: filepath.Join(k.root, ksmPagesToScan),
+	}
+
+	k.sleepInterval = sysfsAttribute{
+		path: filepath.Join(k.root, ksmSleepMillisec),
+	}
+
+	k.run = sysfsAttribute{
+		path: filepath.Join(k.root, ksmRunFile),
+	}
+
+	if err := k.run.open(); err != nil {
+		return err
+	}
+
+	if err := k.sleepInterval.open(); err != nil {
+		_ = k.run.close()
+		return err
+	}
+
+	if err := k.pagesToScan.open(); err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialPagesToScan, err = k.pagesToScan.read()
+	if err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialSleepInterval, err = k.sleepInterval.read()
+	if err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialKSMRun, err = k.run.read()
+	if err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialized = true
+	k.kickChannel = make(chan bool)
+
+	return nil
+}
+
+// restoreSysFS is unlocked. You should take the ksm lock before calling it.
+func (k *ksm) restoreSysFS() error {
+	var err error
+
+	if !k.initialized {
+		return errors.New("KSM is unavailable")
```

Comment:
This error is used multiple times so could be defined as a global to avoid duplication.